core/store/src/migrations.rs

```diff
@@ -512,14 +514,103 @@ pub fn migrate_14_to_15(path: &String) {
     set_store_version(&store, 15);
 }
 
+pub fn migrate_16_to_17(path: &String) {
+    use std::convert::TryFrom;
+
+    use near_primitives::challenge::SlashedValidator;
+    use near_primitives::types::{Balance, BlockHeight, EpochId, ValidatorStake};
+    use near_primitives::utils::index_to_bytes;
+    use near_primitives::version::ProtocolVersion;
+
+    // Migrate from OldBlockInfo to NewBlockInfo - add hash
+    #[derive(BorshDeserialize)]
+    struct OldBlockInfo {
+        pub height: BlockHeight,
+        pub last_finalized_height: BlockHeight,
+        pub last_final_block_hash: CryptoHash,
+        pub prev_hash: CryptoHash,
+        pub epoch_first_block: CryptoHash,
+        pub epoch_id: EpochId,
+        pub proposals: Vec<ValidatorStake>,
+        pub validator_mask: Vec<bool>,
+        pub latest_protocol_version: ProtocolVersion,
+        pub slashed: Vec<SlashedValidator>,
+        pub total_supply: Balance,
+    }
+    #[derive(BorshSerialize)]
+    struct NewBlockInfo {
+        pub hash: CryptoHash,
+        pub height: BlockHeight,
+        pub last_finalized_height: BlockHeight,
+        pub last_final_block_hash: CryptoHash,
+        pub prev_hash: CryptoHash,
+        pub epoch_first_block: CryptoHash,
+        pub epoch_id: EpochId,
+        pub proposals: Vec<ValidatorStake>,
+        pub validator_mask: Vec<bool>,
+        pub latest_protocol_version: ProtocolVersion,
+        pub slashed: Vec<SlashedValidator>,
+        pub total_supply: Balance,
+    }
+    let store = create_store(path);
+    map_col_from_key(&store, DBCol::ColBlockInfo, |key| {
+        let hash = CryptoHash::try_from(key).unwrap();
+        let old_block_info =
+            store.get_ser::<OldBlockInfo>(DBCol::ColBlockInfo, key).unwrap().unwrap();
+        NewBlockInfo {
+            hash,
+            height: old_block_info.height,
+            last_finalized_height: old_block_info.last_finalized_height,
+            last_final_block_hash: old_block_info.last_final_block_hash,
+            prev_hash: old_block_info.prev_hash,
+            epoch_first_block: old_block_info.epoch_first_block,
+            epoch_id: old_block_info.epoch_id,
+            proposals: old_block_info.proposals,
+            validator_mask: old_block_info.validator_mask,
+            latest_protocol_version: old_block_info.latest_protocol_version,
+            slashed: old_block_info.slashed,
+            total_supply: old_block_info.total_supply,
+        }
+    })
+    .unwrap();
+
+    // Add ColHeaderHashesByHeight
+    let chunk_tail =
+        store.get_ser::<BlockHeight>(ColBlockMisc, CHUNK_TAIL_KEY).unwrap().unwrap_or(0);
```

Comment:
shouldn't it be unwrap_or(genesis_height)?

Reply:
This is a corner case for very small tests.

BTW, how is it possible to store Headers with Height lower than genesis_height?

Reply:
I see. I think I misunderstood how it works